### ISSUE TITLE
deferUnlambda: False positive for mutated variable

### ISSUE THREAD

**Issue by ajscholl** [go-critic/go-critic#1401]
```golang
func example() {
	foo := func() {
		println("should never print")
	}
	defer foo()
	defer func() { // deferUnlambda: can rewrite as `defer foo()` (gocritic)
		foo()
	}()
	foo = func() {
		println("should print")
	}
}
```
Following the suggestion changes the behavior of the program. Instead of calling whatever `foo` was set to at the end of the function, we now instead call whatever `foo` was set to at the time of the `defer` call.

**Comment by ehsundar** [comment]
Need to convert this rule to a `StmtListVisitor` to mitigate this issue. Do you agree?  @cristaloleg

**Comment by cristaloleg** [comment]
For now sounds ok. Feel free to try.
